src/cache/cache/green_cache.cpp

```diff
@@ -0,0 +1,181 @@
+// Rafael Sá Menezes - 04/2020
+
+#include "green_cache.h"
+#include <cache/algorithms/algorithms.h>
+#include <cache/algorithms/algorithms_utils.h>
+
+void green_cache::canonize_expr(expr2tc &expr)
+{
+  expr_variable_reordering reordering(expr);
+  reordering.run();
+
+  expr_green_normal_form gnf(expr);
+  gnf.run();
+
+  // TODO: add variable renaming
+}
+
+void green_cache::run_on_assert(symex_target_equationt::SSA_stept &step)
+{
+  expr2tc &cond = step.cond;
+
+  // First assert irep should begin with an implies
+  assert(cond->expr_id == expr2t::expr_ids::implies_id);
+
+  std::shared_ptr<logic_2ops> implies;
+  implies = std::dynamic_pointer_cast<logic_2ops>(cond);
+
+  expr2tc &lhs(implies->side_1);
+  std::shared_ptr<symbol_data> symbol;
+  symbol = std::dynamic_pointer_cast<symbol_data>(lhs);
+  std::string guard_name = symbol->get_symbol_name();
+
+  expr2tc &rhs(implies->side_2);
+
+  crc_expr guard_items;
+
+  // If the assertive contains inner guards then rhs will be an imply
+  if(rhs->expr_id == expr2t::expr_ids::implies_id)
+  {
+    /*
+     * If an ASSERTIVE contains an inner implies this means that this is in the
+     * format:
+     *
+     * GUARD0 && GUARD1 && ... && GUARDN => ¬inner_expr
+     *
+     * LHS does not need to be canonized and can contain || besides &&
+     * TODO: Add support to ||
+     */
+    std::shared_ptr<logic_2ops> inner_implies;
+    inner_implies = std::dynamic_pointer_cast<logic_2ops>(rhs);
+
+    // Canonize inner_expr
+    canonize_expr(inner_implies->side_2);
+    auto inner_expr_hash = convert_expr_to_hash(inner_implies->side_2);
+    guard_items.insert(inner_expr_hash);
+    parse_implication_guard(inner_implies->side_1, guard_items);
+  }
+  else
+  {
+    // Simple guard
+    canonize_expr(rhs);
+    auto expr_hash = convert_expr_to_hash(rhs);
+    guard_items.insert(expr_hash);
+  }
+
+  items[guard_name] = guard_items;
+
+  load_unsat_container();
+  if(unsat_container.check(guard_items))
+  {
+    /**
+     * An assertive is in the format of
+     *
+     * a. assertive_guard -> !expr
+     * b. assertive_guard -> guards && ... && guard -> !expr.
+     *
+     * For 'a' the is the trivial case, if expr is known to be false
+     * then we mark it's negation as true.
+     *
+     * In 'b', using basic logic equivalence:
+     *
+     * A -> B <-> ¬A OR B
+     *
+     * Where A would be the set of guards and B the !expr. Since A if false
+     * it's negation is true, so we can simplify to:
+     *
+     * assertive_guard -> 1 (true)
+     *
+     */
+    constant_bool2tc false_value(true);
+    step.cond = false_value;
+  }
+}
+
+void green_cache::parse_implication_guard(
```

Comment:
Add short description for all methods.

Reply:
The header includes the description of the method